planner/core/task.go

```diff
@@ -126,46 +138,139 @@ func (p *PhysicalApply) attach2Task(tasks ...task) task {
+	outerCnt, innerCnt := outerTask.count(), innerTask.count()
+	// Add the cost of evaluating outer filter, since inner filter of index join
+	// is always empty, we can simply tell whether outer filter is empty using the
+	// summed length of left/right conditions.
+	if len(p.LeftConditions)+len(p.RightConditions) > 0 {
+		cpuCost += cpuFactor * outerCnt
+		outerCnt *= selectionFactor
+	}
+	// Cost of extracting lookup keys.
+	innerCPUCost := cpuFactor * outerCnt
+	// Cost of sorting and removing duplicate lookup keys:
+	// (outerCnt / batchSize) * (batchSize * Log2(batchSize) + batchSize) * cpuFactor
+	batchSize := math.Min(float64(p.ctx.GetSessionVars().IndexJoinBatchSize), outerCnt)
+	if batchSize > 2 {
+		innerCPUCost += outerCnt * (math.Log2(float64(batchSize)) + 1) * cpuFactor
+	}
+	// Add cost of building inner executors. CPU cost of building copTasks:
+	// (outerCnt / batchSize) * (batchSize * distinctFactor) * cpuFactor
+	// Since we don't know the number of copTasks built, ignore these network cost now.
+	innerCPUCost += outerCnt * distinctFactor * cpuFactor
+	// CPU cost of building hash table for inner results:
+	// (outerCnt / batchSize) * (batchSize * distinctFactor) * innerCnt * cpuFactor
+	innerCPUCost += outerCnt * distinctFactor * innerCnt * cpuFactor
+	innerConcurrency := float64(p.ctx.GetSessionVars().IndexLookupJoinConcurrency)
+	cpuCost += innerCPUCost / innerConcurrency
+	// Cost of probing hash table in main thread.
+	numPairs := outerCnt * innerCnt
+	if p.JoinType == SemiJoin || p.JoinType == AntiSemiJoin ||
+		p.JoinType == LeftOuterSemiJoin || p.JoinType == AntiLeftOuterSemiJoin {
+		if len(p.OtherConditions) > 0 {
+			numPairs *= 0.5
+		} else {
+			numPairs = 0
+		}
 	}
-	cst := lCnt * netWorkFactor
-	batchSize := p.ctx.GetSessionVars().IndexJoinBatchSize
-	cst += lCnt * math.Log2(math.Min(float64(batchSize), lCnt)) * 2
-	cst += lCnt / float64(batchSize) * netWorkStartFactor
-	return cst
+	probeCost := numPairs * cpuFactor
+	// Cost of additional concurrent goroutines.
+	cpuCost += probeCost + (innerConcurrency+1.0)*concurrencyFactor
+	// Memory cost of hash tables for inner rows. The computed result is the upper bound,
+	// since the executor is pipelined and not all workers are always in full load.
+	memoryCost := innerConcurrency * (batchSize * distinctFactor) * innerCnt * memoryFactor
+	// Cost of inner child plan, i.e, mainly I/O and network cost.
+	innerPlanCost := outerCnt * innerTask.cost()
+	return outerTask.cost() + innerPlanCost + cpuCost + memoryCost
 }
 
-func (p *PhysicalHashJoin) getCost(lCnt, rCnt float64) float64 {
-	smallTableCnt := lCnt
+// GetCost computes cost of hash join operator itself.
+func (p *PhysicalHashJoin) GetCost(lCnt, rCnt float64) float64 {
+	innerCnt, outerCnt := lCnt, rCnt
 	if p.InnerChildIdx == 1 {
-		smallTableCnt = rCnt
+		innerCnt, outerCnt = rCnt, lCnt
+	}
+	// Cost of building hash table.
+	cpuCost := innerCnt * cpuFactor
+	memoryCost := innerCnt * memoryFactor
```

Comment:
Also, we encode the inner join keys into bytes. Maybe we also need to consider memory consumption of this.